google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,288 @@
+# Copyright 2017 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Google
+  module Cloud
+    module Spanner
+      ##
+      # # Interval
+      #
+      # Represents an interval of time by storing the time components
+      # in months, days and nanoseconds.
+      #
+      # @example
+      #   require "google/cloud/spanner"
+      #
+      #   iso_8601_string = "P1Y2M3DT4H5M6S"
+      #   interval = Google::Cloud::Spanner::Interval::parse iso_8601_string
+      #
+      #   puts interval # "P1Y2M3DT4H5M6S"
+      class Interval
+        NANOSECONDS_IN_A_SECOND = 1_000_000_000
+        NANOSECONDS_IN_A_MINUTE = NANOSECONDS_IN_A_SECOND * 60
+        NANOSECONDS_IN_AN_HOUR = NANOSECONDS_IN_A_MINUTE * 60
+        NANOSECONDS_IN_A_MILLISECOND = 1_000_000
+        NANOSECONDS_IN_A_MICROSECOND = 1_000
+        MAX_MONTHS = 120_000
+        MIN_MONTHS = -MAX_MONTHS
+        MAX_DAYS = 3_660_000
+        MIN_DAYS = -MAX_DAYS
+        MAX_NANOSECONDS = 316_224_000_000_000_000_000
+        MIN_NANOSECONDS = -316_224_000_000_000_000_000
+
+        private_constant :NANOSECONDS_IN_A_SECOND, :NANOSECONDS_IN_A_MINUTE, :NANOSECONDS_IN_AN_HOUR,
+                         :NANOSECONDS_IN_A_MILLISECOND, :NANOSECONDS_IN_A_MICROSECOND, :MAX_MONTHS,
+                         :MIN_MONTHS, :MAX_DAYS, :MIN_DAYS, :MAX_NANOSECONDS, :MIN_NANOSECONDS
+
+        class << self
+          # Parses an ISO8601 string and returns an Interval instance.
+          #
+          # The accepted format for the ISO8601 standard is:
+          # `P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S`
+          # where `n` represents an integer number.
+          #
+          # @param interval_string [String] An ISO8601 formatted string.
+          # @return [Google::Cloud::Spanner::Interval]
+          #
+          # @example
+          #   require "google/cloud/spanner"
+          #
+          #   iso_8601_string = "P1Y2M3DT4H5M6S"
+          #   interval = Google::Cloud::Spanner::Interval::parse iso_8601_string
+          #
+          #   puts interval # "P1Y2M3DT4H5M6S"
+          def parse interval_string
+            pattern = /^
+              P(?!$)(?:(?<years>-?\d+)Y)?(?:(?<months>-?\d+)M)?(?:(?<days>-?\d+)D)
+              ?(?:T(?!$)(?:(?<hours>-?\d+)H)?(?:(?<minutes>-?\d+)M)?(?:(?<seconds>-?(?!S)\d*(?:[\.,]\d{1,9})?)S)?)?
+              $
+            /x
```

Comment:
If you're going to use the `x` modifier, you might as well make this a lot more readable. For example, put each section (e.g. number-followed-by-letter) on its own line. Right now the break between lines 67 and 68 separates a section from the `?` that modifies it, which is probably not a great place to put a line break.